Merge daemon defaults with user-supplied jvmargs

bigdaz opened the pull request:
Fixes #19750

### Context

Previously, setting any value for `org.gradle.jvmargs` caused all default settings to be lost.
This often resulted in important defaults like `-XXMaxMetaspaceSize` being omitted when a user
attempted to provide more memory to a build process.

With this change, default jvmargs will be retained unless specifically overridden by a user-supplied argument.
One exception is that setting either -Xmx or -Xms will cause the default heap size settings to be omitted, preventing
user-supplied values from conflicting with default values (like having a min heap larger than max heap).

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

bigdaz committed:
Merge default with user-supplied jvmargs

Previously, setting any value for `org.gradle.jvmargs` caused all default settings to be lost.
This often resulted in important defaults like `-XXMaxMetaspaceSize` being omitted when a user
attempted to provide more memory to a build process.

With this change, default jvmargs will be retained unless specifically overridden by a user-supplied argument.
One exception is that setting either -Xmx or -Xms will cause the default heap size settings to be omitted, preventing
user-supplied values from conflicting with default values (like having a min heap larger than max heap).

Fixes #19750

bigdaz committed:
Document correct defaults for Daemon JVM

bigdaz committed:
Remove redundant support for daemon on JDK < 8

bigdaz commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"cace570a2c9df80b329d7fe5043f83ff52a3a0dd","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered the following builds for you:
- [PullRequestFeedback build](https://builds.gradle.org/viewQueued.html?itemId=49964316)

bigdaz committed:
Remove minimum JDK check from DaemonParameters

The minimum JDK is verified elsewhere, and this failure interferes
with reporting.

bigdaz commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"181270c585063b59235146f655d1e53b7347bbc8","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered the following builds for you:
- [PullRequestFeedback build](https://builds.gradle.org/viewQueued.html?itemId=50010836)

bigdaz committed:
Fix test that was assuming JVM arg override

bigdaz commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"3e74b29f9c5352deda1e129d0ac5dedb25d8e14c","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered the following builds for you:
- [PullRequestFeedback build](https://builds.gradle.org/viewQueued.html?itemId=50161965)

bigdaz commented:
@bot-gradle test this

bigdaz commented:
@bot-gradle test this

bigdaz commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"1475cff63d54198be917a1bec3f2ab787ea51a09","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered the following builds for you:
- [PullRequestFeedback build](https://builds.gradle.org/viewQueued.html?itemId=50461848)

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"1475cff63d54198be917a1bec3f2ab787ea51a09","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered the following builds for you:
- [PullRequestFeedback build](https://builds.gradle.org/viewQueued.html?itemId=50462045)

octylFractal reviewed (changes requested):
Thanks for your contribution. I think the change looks good overall, but I would like some more docs covering the new merging behavior.

bigdaz commented:
@octylFractal Thanks for the feedback. Unfortunately this PR is hitting some [hard-to-understand test failures](https://ge.gradle.org/s/tpud6st67opl4/tests/overview?outcome=failed) that will require further investigation before this change could be merged. I'm not sure if/when I'll find time to do that.

In the meantime, would it be helpful if I close this PR (or convert to draft?)

octylFractal commented:
I'm looking at the test failures now, so don't worry about that.

bigdaz committed:
Improve docs for `org.gradle.jvmargs` defaults

bigdaz commented on subprojects/docs/src/docs/userguide/running-builds/build_environment.adoc:
```diff
@@ -96,7 +96,7 @@ _A reasonable default is derived from your environment (`JAVA_HOME` or the path
 
 `org.gradle.jvmargs=(JVM arguments)`::
 Specifies the JVM arguments used for the Gradle Daemon. The setting is particularly useful for <<#sec:configuring_jvm_memory,configuring JVM memory settings>> for build performance. This does not affect the JVM settings for the Gradle client VM.
```
I've pushed some improvements to the doc. PTAL: 438a568

octylFractal replied:
I made the wording better (in my opinion), PTAL: 24a222e8a36f751b598c0f60150af0d27bf9c008

octylFractal committed:
Improve wording of JVM arg docs

octylFractal commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"24a222e8a36f751b598c0f60150af0d27bf9c008","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered the following builds for you:
- [PullRequestFeedback build](https://builds.gradle.org/viewQueued.html?itemId=50917803)

octylFractal commented:
@bigdaz I looked into the "hard-to-understand test failures", and the underlying issue is that we used to let anything that set `org.gradle.jvmargs` run with unlimited `Metaspace` if not specified, and now we limit it to `256m`. After some research, @DPUkyle and I came to the conclusion that it would be a good idea to increase this to perhaps `1G` by default, or remove it entirely. What do you think?

bigdaz commented:
We should not remove the setting altogether. The entire reason for this PR is that users are setting `org.gradle.jvmargs` _without_ specifying `MaxMetaspaceSize`, and the daemon process is consuming more and more memory until the process dies.
See #19750 for details.

Providing a higher default value might make sense. Or since these failures seem specific to the Kotlin compiler daemon (I think), perhaps the best fix is to ensure an appropriate `MaxMetaspaceSize` for this process.

bigdaz commented:
@eskatos I see you assigned this back to me, but I'm not sure what action can/should be taken.
The current behaviour is that if you don't set _any_ jvm args, then we set `MaxMetaspaceSize=256m`. If a user sets a completely unrelated jvm arg, then we don't provide any default value for `MaxMetaspaceSize`.

Perhaps this makes sense: we provide a default _set_ of jvm args and you can choose to override the entire set, but not one value in the set. But I've seen a number of users struggle with this: they give their build _more_ memory (say `-Xmx1024m`), and their build starts to run out of memory! It's a slightly different out of memory error, but users don't really know how to fix this.

There are a few options to address this:
1. Do nothing, but perhaps improve the documentation so it's clear that setting `-Xmx` does more than just giving more memory.
2. When a user sets just one of the values that we provide a default for, we only change _that_ value, and leave the others in place. (That's what my PR tries to do).
3. Provide a different (or higher) default value for `MaxMetaspaceSize`. We should probably document this. 
4. Provide a better model for specifying the memory settings for a build, allowing users to set value independently.
5. ???

I could help out with any of 1-3. I think 4 would require more work and should be tackled by the BT team.
WDYT?

eskatos commented:
I assigned it back to you while triaging unassigned PRs and adding assignees to all team members/authors.

I think this PR is a reasonable improvement and will address the current confusion. But as shown by the failing tests, this may break builds.

I'm not sure if raising the default metaspace size to 1g would be a good move. Loots of builds won't need that much.

@octylFractal, @big-guy, would the approaching 8.0 be a good time to address this? Can you take over making a decision on this?

jvandort committed:
Fix setting of expanded JVM memory arguments

The existing implementation was overwriting the jvm args getting set in AbstractSmokeTest. This
caused the default memory args to get set, which are much lower than the default memory
settings for an un-configured jvm.

jvandort commented:
@bot-gradle test ReadyForNightly

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"47972d8b06dcc697dce6699e0ffe9a62add6501e","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered the following builds for you:
- [ReadyForNightly build](https://builds.gradle.org/viewQueued.html?itemId=54357651)

jvandort committed:
Add upgrade guide entry

jvandort committed:
Fix DaemonParametersTest by fixing arg ordering

jvandort commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"d66c22b3ad50afca230dc17ae8f62bf1ff37b79f","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered the following builds for you:
- [PullRequestFeedback build](https://builds.gradle.org/viewQueued.html?itemId=54380149)

big-guy committed:
Merge remote-tracking branch 'origin/master' into dd/issue/19750

* origin/master: (155 commits)
  Clarify docs wording for type-safe deps API
  Correct vcs root id for release/experimental branches
  Add placeholders for 8.0 upgrade guide
  Downgrade to version which doesn't require Java 11
  Update the version of the Grgit plugin used in a test
  Revert changes to TestNG default version
  Update released version to latest snapshot
  Update library versions in build init to latest for 7.7
  Clean release notes and welcome message for 7.7
  Update version to 7.7
  Clean accepted API changes
  Update wrapper to latest nightly
  Not show build scan link for `updateReleasedVersions` task
  Update to 7.5.1
  Publish 7.5.1-20220805205215+0000
  Refreeze arch unit expectations
  Finalize release notes for patch release
  Publish 7.5.1-20220805123425+0000
  Rework test to demonstrate usage of module-source-path
  Refresh released version list
  ...

big-guy committed:
Move message about potential breaking changes to jvmargs handling to 8.0

big-guy commented:
@bot-gradle test and merge

bot-gradle commented:
OK, I've already triggered [a build](https://builds.gradle.org/viewQueued.html?itemId=54815160) for you.

bot-gradle commented:
[Pre-tested commit build](https://builds.gradle.org/viewLog.html?buildId=54815160) failed.

blindpirate commented:
The performance test asserts there's only one daemon involved in multiple iterations. However, this PR seems to change something in daemon compatibility so there're multiple daemon started. The performance test failure is reproducbile.

nathan-contino commented on subprojects/docs/src/docs/userguide/migration/upgrading_version_7.adoc:
```diff
@@ -46,7 +46,13 @@ TBD
 === Potential breaking changes
 
-TBD
+==== User provided daemon JVM arguments are merged with defaults
+
+Prior to 8.0, setting `org.gradle.jvmargs` would overwrite all default daemon JVM args, however now Gradle's default JVM arguments are merged with the arguments provided by this property. In some cases, this may cause builds which set `org.gradle.jvmargs` to have different memory configuration in Gradle 7.6.
+
+As it was before, the default daemon JVM args are `-Xms256m -Xmx512m -XX:MaxMetaspaceSize=256m -XX:+HeapDumpOnOutOfMemoryError`. Previously, if `org.gradle.jvmargs` is set to `-DsomeProperty=123`, the existing behavior would be to resolve the daemon arguments to `-DsomeProperty=123` while the new behavior would resolve to `-DsomeProperty=123 -Xms256m -Xmx512m -XX:MaxMetaspaceSize=256m -XX:+HeapDu ...
+
+Since the JVM's default memory arguments are more generous than Gradle's default daemon JVM arguments, this change can cause existing builds to fail with an `OutOfMemoryError` if the build uses more memory than is explicitly set in `org.gradle.jvmargs`.
```
```suggestion
==== User provided daemon JVM arguments merge with defaults

Prior to 8.0, setting `org.gradle.jvmargs` overwrites all default daemon JVM args. In 8.0 and above, Gradle merges default JVM arguments with the arguments provided by this property. This can effect the memory configuration of builds that set `org.gradle.jvmargs`.

Consider the default daemon JVM arguments:

----
-Xms256m -Xmx512m -XX:MaxMetaspaceSize=256m -XX:+HeapDumpOnOutOfMemoryError
----

Prior to 8.0, if a user assigned `org.gradle.jvmargs` the value `-DsomeProperty=123`, Gradle resolved the daemon arguments to the following:

----
-DsomeProperty=123
----

In 8.0 and above, if a user assigns `org.gradle.jvmargs` the value `-DsomeProperty=123`, Gradle resolves the daemon arguments to the following:

----
-DsomeProperty=123 -Xms256m -Xmx512m -XX:MaxMetaspaceSize=256m -XX:+HeapDumpOnOutOfMemoryError`
----

Since the JVM's default memory arguments are more generous than Gradle's default daemon JVM arguments, this change can cause existing builds to fail with an `OutOfMemoryError` if the build uses more memory than explicitly set in `org.gradle.jvmargs`.
```

Suggestion: use the terms "Before 8.0" and "In 8.0 and above" to make the subject of sentences (and comparisons) clearer.

Separate large collections of JVM arguments into block code snippets to make them more readable.

Use active voice when possible to make it clearer when Gradle behaves in some way, and when the user behaves in some way. I find the docs a little hard to parse as written.

nathan-contino commented on subprojects/docs/src/docs/userguide/running-builds/build_environment.adoc:
```diff
@@ -100,8 +100,9 @@ Specifies the Java home for the Gradle build process. The value can be set to ei
 `org.gradle.jvmargs=(JVM arguments)`::
-Specifies the JVM arguments used for the Gradle Daemon. The setting is particularly useful for <<#sec:configuring_jvm_memory,configuring JVM memory settings>> for build performance. This does not affect the JVM settings for the Gradle client VM.
-_The default is `-Xmx512m "-XX:MaxMetaspaceSize=256m"`._
+Specifies JVM arguments to be used for the Gradle Daemon. The setting is particularly useful for <<#sec:configuring_jvm_memory,configuring JVM memory settings>> for build performance. This does not affect the JVM settings for the Gradle client VM.
+Supplied values are merged with the defaults, with special handling for <<#sec:configuring_jvm_memory,JVM memory settings>> (see link).
+_The default values are `-Xms256m -Xmx512m -XX:MaxMetaspaceSize=256m -XX:+HeapDumpOnOutOfMemoryError`._
```
```suggestion
Specifies the JVM arguments to use for the Gradle Daemon. The setting is particularly useful for <<#sec:configuring_jvm_memory,configuring JVM memory settings>> for build performance. This does not affect the JVM settings for the Gradle client VM.
Gradle merges supplied values with the defaults, with special handling for <<#sec:configuring_jvm_memory,JVM memory settings>>.
_The default value is `-Xms256m -Xmx512m -XX:MaxMetaspaceSize=256m -XX:+HeapDumpOnOutOfMemoryError`._
```
Suggestion: slight reword for docs consistency. I suspect we should specify the default *value* here (not values) because the default value refers to the singular `org.gradle.jvmargs` property, not the plural JVM arguments.

nathan-contino commented on subprojects/docs/src/docs/userguide/running-builds/build_environment.adoc:
```diff
@@ -325,11 +326,14 @@ You should check for existence of optional project properties before you access
 You can adjust JVM options for Gradle in the following ways:
 
-The `org.gradle.jvmargs` Gradle property controls the VM running the build. It defaults to `-Xmx512m "-XX:MaxMetaspaceSize=256m"`
+The `org.gradle.jvmargs` Gradle property specifies arguments for the JVM running the build. The default values are `-Xms256m -Xmx512m -XX:MaxMetaspaceSize=256m -XX:+HeapDumpOnOutOfMemoryError`.
```
```suggestion
The `org.gradle.jvmargs` Gradle property specifies arguments for the JVM running the build. The default value is `-Xms256m -Xmx512m -XX:MaxMetaspaceSize=256m -XX:+HeapDumpOnOutOfMemoryError`.
```
Same suggestion as above -- I *think* this ought to be singular.

nathan-contino commented on subprojects/docs/src/docs/userguide/running-builds/build_environment.adoc:
```diff
@@ -325,11 +326,14 @@ You should check for existence of optional project properties before you access
+The `org.gradle.jvmargs` Gradle property specifies arguments for the JVM running the build. The default values are `-Xms256m -Xmx512m -XX:MaxMetaspaceSize=256m -XX:+HeapDumpOnOutOfMemoryError`.
 
-.Changing JVM settings for the build VM
+JVM arguments provided via `org.gradle.jvmargs` are combined with the defaults, with special handling for `-Xms` and `-Xmx`: setting either these values will remove the default for the other.
+This avoids the case where a default value is incompatible with a user-supplied value, such as having a minimum memory setting higher than the maximum.
```
```suggestion
Gradle combines JVM arguments provided via `org.gradle.jvmargs` with the default JVM arguments. Setting either `-Xms` and `-Xmx` removes the default for the other.
This avoids incompatible combinations of default and user-supplied values, such as a minimum memory setting higher than the maximum.
```
Suggestion: active voice rephrase to make the first sentence more readable. Trimmed some words from the second sentence for the same reason.

nathan-contino reviewed (approved):
Docs LGTM, a couple of optional suggestions

big-guy commented:
We're not able to look at this until 2025

runningcode commented:
Bump. This leads to so much frustration with Gradle users who aren’t aware of this footgun.

bigdaz commented:
@big-guy I understand that this was considered a breaking change, but can we please get the behaviour changed for Gradle 9.0?
